src/NuGet.Core/NuGet.PackageManagement/Resolution/ResolverGather.cs

```diff
@@ -502,69 +523,83 @@ private void QueueWork(IReadOnlyList<SourceResource> sources, PackageIdentity pa
+                // a resource may be null, if it is exclude this source from the gather
+                foreach (var source in _context.PrimarySources)
                 {
-                    var resource = await depResources[source];
-
-                    if (source != null && !_primaryResources.Any(sourceResource => sourceResource.Source.PackageSource.Equals(source)))
+                    if (uniquePrimarySources.Add(source.PackageSource))
                     {
-                        _primaryResources.Add(new SourceResource(source, resource));
+                        var resource = await depResources[source];
+
+                        if (source != null && !_primaryResources.Any(sourceResource => sourceResource.Source.PackageSource.Equals(source)))
+                        {
+                            _primaryResources.Add(new SourceResource(source, resource));
+                        }
                     }
                 }
-            }
 
-            // All sources - for fallback
-            var uniqueAllSources = new HashSet<Configuration.PackageSource>();
+                // All sources - for fallback
+                var uniqueAllSources = new HashSet<Configuration.PackageSource>();
 
-            foreach (var source in allSources)
-            {
-                if (uniqueAllSources.Add(source.PackageSource))
+                foreach (var source in allSources)
                 {
-                    var resource = await depResources[source];
-
-                    if (source != null && !_allResources.Any(sourceResource => sourceResource.Source.PackageSource.Equals(source)))
+                    if (uniqueAllSources.Add(source.PackageSource))
                     {
-                        _allResources.Add(new SourceResource(source, resource));
+                        //var resource = await depResources[source];
+                        var resource = depResources[source];
+
+                        if (source != null && !_allResources.Any(sourceResource => sourceResource.Source.PackageSource.Equals(source)))
+                        {
+                            currentSource = source.PackageSource.Source;
+                            _allResources.Add(new SourceResource(source, resource.Result));
+                        }
                     }
                 }
+
+
+                // Installed packages resource
+                _packagesFolderResource = await _context.PackagesFolderSource.GetResourceAsync<DependencyInfoResource>(token);
+            }
+            catch (Exception ex) when (ex is System.Net.Http.HttpRequestException || ex is OperationCanceledException ||
+                                       ex is InvalidOperationException || ex is TaskCanceledException || ex is AggregateException)
+            {
+                string message = String.Format(Strings.ExceptionWhenTryingToAddSource, ex.GetType().ToString(), currentSource);
+                throw new InvalidOperationException(message);
```

Comment:
Can you set the InnerException on this exception? so...
throw new InvalidOperationException(message, ex);